### ISSUE TITLE
[Feature request] DefaultTabController.onPageChanged

### ISSUE THREAD

**Issue by mtc-jed** [flutter/flutter#101690]
## Use case

I'd like to update some data when the user changes Tab in a TabBarView. Since DefaultTabController's internal controller isn't accessible, I can't listen for changes and have to setup my own controller.

## Proposal
Add a `void Function(int)? onPageChanged` argument to `DefaultTabController` which would be called whenever the user changes Tabs (passing the current index to the provided callback).

**Comment by chunhtai** [comment]
I think this can be fix by provide a different constructor of DefaultTabController that takes in a TabController instead.

**Comment by Srivatsa10** [comment]
What data do you want to change?

**Comment by chunhtai** [comment]
@mtc-jed 

Yes I think that mostly makes sense. I believe there are two benefit of DefaultTabController

1. no need to have vsync to create tab controller
2. no need to manually pass-in the tab controller to the tabview in the subtree.

If DefaultTabController takes in a tabcontroller, we lose benefit 1, but we still have benefit 2

I don't like the API of `DefaultTabController.onPageChanged` because we may end up adding more and more APIs to listen/control to the tabcontroller and makes it complex and hard to maintain.

Side node: The code 
```dart
final TabController controller = DefaultTabController.of(context);
controller.addListener(() {
  if (!controller.indexIsChanging)
    setTitle();
});
```
should be in didChangeDependency, not build method since you are modifying the state of the inherited object.

**Comment by mtc-jed** [comment]
> I don't like the API of `DefaultTabController.onPageChanged` because we may end up adding more and more APIs to listen/control to the tabcontroller and makes it complex and hard to maintain.

@chunhtai Indeed, is seems dumb to "double" every API (and keep doing so in the future). Could it be possible to make the vsync argument of the tab controller nullable ? In which case, we'd be able to create it ourselves and pass it to DefaultTabController, which would put its own vsync.
There's of course still the problem that `_DefaultTabControllerState._controller` get re-created, and I don't think making vsync nullable is a good idea in general.

`DefaultTabController.onPageChanged` really is about convenience. It all depends on how much convenience we're willing to give the end user, at the cost of complexity in the framework. This API seems like a good, almost cost-free convenience to me, but that _really_ is a subjective thing.
I'll probably end up making my own "convenience widget", as you regularly do.

**Comment by vulpeep** [comment]
> Shouldn't this behavior be avoided or did i miss something? Just run the example above, make 2-3 hot reloads and then change the tabs. The print statement would be called multiple times.

As a temporary solution, I'm using a check for existing listeners:
```dart
              if (!controller.hasListeners) {
                controller.addListener(() {
                  if (!controller.indexIsChanging)
                    print(controller.index);
                });
              }
```
